Measuring program progress

rlucioni opened the pull request:
Introduces a utility class for gauging a user's progress towards program completion. Progress data is passed to the ProgramListFactory. ECOM-3200.

@jimabramson @schenedx, please review. @AlasdairSwan FYI, I think you'll be most interested in my addition to `lms/templates/learner_dashboard/programs.html`.

schenedx commented on openedx/core/djangoapps/programs/tests/mixins.py:
```diff
@@ -33,7 +33,10 @@ def create_programs_config(self, **kwargs):
-    """Mixin mocking Programs API URLs and providing fake data for testing."""
+    """Mixin mocking Programs API URLs and providing fake data for testing.
+
+    NOTE: This mixin is DEPRECATED. Tests should create and manage their own data.
```
If this is deprecated, why not just go ahead and remove it?

rlucioni replied:
@schenedx removing this is going to require retrofitting many tests that currently depend on it. Making those changes now would increase the scope of this PR significantly. I intend to remove the mixin in a separate PR.

schenedx commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -132,28 +133,120 @@ def get_display_category(program):
 
-def get_engaged_programs(user, enrollments):
-    """Derive a list of programs in which the given user is engaged.
+class ProgramProgressMeter(object):
```
I don't like the word "meter". This is a class to calculate the user's program progress. How about "ProgramProgressCalculator"?

schenedx commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -132,28 +133,120 @@ def get_display_category(program):
+                towards completing a program.
+        """
+        if self.engaged_programs is None:
+            self.programs  # pylint: disable=pointless-statement
```
I think it's better to avoid silly statements like this. What about create a program initialization method and call that in the constructor? Then the property can be already populated and used.

jimabramson replied:
can i ask why you didn't just do `pass` here?

jimabramson replied:
or is it a typo maybe?  if `engaged_programs` is `None` we're headed for a big fat Exception on L184.

rlucioni replied:
I think my use of `@property` on the `programs` method might be causing confusion here. This check is intended to make sure that `engaged_programs` is populated before we proceed with progress calculation. I'll take @schenedx's suggestion and populate `engaged_programs` during initialization.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_utils.py:
```diff
@@ -10,15 +11,18 @@
 from provider.constants import CONFIDENTIAL
 
+from certificates.models import CertificateStatuses, GeneratedCertificate  # pylint: disable=import-error
+from lms.djangoapps.certificates.tests.factories import GeneratedCertificateFactory
```
same djangoapp - why the different module addressing styles?

rlucioni replied:
To get around an error about conflicting models documented [here](https://code.djangoproject.com/ticket/22280). Importing the models from `lms.djangoapps.certificates.models` raises a `RuntimeError`. I don't like it either, but I'm having a hard time understanding how to fix the problem. It looks like this isn't a new issue, either: https://github.com/edx/edx-platform/search?l=python&q=%22certificates.models%22&utf8=%E2%9C%93.

schenedx commented on openedx/core/djangoapps/programs/tasks/v1/tasks.py:
```diff
@@ -37,6 +36,7 @@ def get_api_client(api_config, student):
     return EdxRestApiClient(api_config.internal_api_url, jwt=id_token)
 
 
+# TODO: Move to program utilities. Will require updating tests.
```
Is it too much to ask to do this refactoring within this PR? I rather not have "utils.py" to depend on this module, let alone the circular dependency

rlucioni replied:
I think doing it as part of this PR would increase the scope unnecessarily. My plan is to do this in a separate PR, to keep things clean and avoid slowing down the more immediately useful parts of this PR. Let me know if you feel strongly about doing it now.

schenedx replied:
Fair point. I feel strongly for not leaving the code in this state, but I agree a separate PR would be cleaner and easier to gain value earlier. Maybe we should not consider the ticket is done until the new refactor PR is merged?

rlucioni replied:
@schenedx I might end up taking care of this now. The way I moved imports around is making it difficult to mock a function in this module during testing.

rlucioni replied:
@schenedx fixed!

jimabramson commented on openedx/core/djangoapps/programs/tests/test_utils.py:
```diff
@@ -191,110 +194,393 @@ def test_get_program_for_certificates_id_not_exist(self):
+                'in_progress': 0,
+                'not_started': 0,
+            }]
+        )
```
this assert is repeated a lot and probably deserves a helper method for shortand, e.g. `self.assert_progress(meter, program['id'], 1, 0, 0)`

rlucioni replied:
@jimabramson good call, I'll look into that.

jimabramson commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -132,28 +133,120 @@ def get_display_category(program):
+                'completed': completed,
+                'in_progress': in_progress,
+                'not_started': not_started,
+            })
```
@rlucioni suggestion.  In the next milestone it will be necessary to display the completed / in progress / not started information correlated with the individual courses/runs.  Instead of returning counts by status in this API,  would it make sense to return tuples of course codes?  One can do `len()` on those values to get the same counts being returned here, but it will also be possible to do the richer displays without API changes.

rlucioni replied:
@jimabramson I take it you mean something like the following?
```python
{
    id': 1,
    'completed': ["A Complete Course"],
    'in_progress': ["An In-Progress Course"],
    'not_started': ["An Unstarted Course", "Another Unstarted Course"],
}
```
I can do this now, although I have questions about approach and how it'll affect ongoing front-end work. Do we want to only provide display names here, or the entire course code object (nested run modes included)?

@AlasdairSwan if I were to do this, would you prefer I pass you counts, as we'd agreed previously, or the raw data and let the front-end do the counting?

jimabramson replied:
@rlucioni my thinking was that the course_code.key would be sufficient given what we need to display down the road.

At the same time, I recognize it would be equally feasible to amend the existing data structure that provides counts, with a second structure that provides the course code details, which would avoid jumping the gun on a story that's not yet in progress.  So, your call here.

jimabramson commented:
@rlucioni apart from inline comments the overall approach LGTM.  One thing, if possible I think it would be more optimal to mock out the certificates api functions than directly populating models via factories - this is work we'll have to undo later when we migrate course certificates to the new credentials service.

rlucioni commented:
@schenedx @jimabramson I've addressed all of your comments. The progress API now returns lists of course code names, and mock is used instead of populating certificate models via factories.

schenedx commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -132,28 +133,142 @@ def get_display_category(program):
+            bool, whether the course code is in progress.
+        """
+        return any([
+            run_mode['course_key'] in self.course_ids
```
In the `_is_complete` function above, the array returned are of format returned by `_parse`. But the format of the return elements here do not use `_parse` function. Why the different item structure? Also, I noticed for not started items, the element returned is a straight up name.

rlucioni replied:
@schenedx neither of these methods returns a list. They use Python's built-in [`any`](https://docs.python.org/3/library/functions.html#any) function to return a Boolean. The `_parse` method is used to allow comparison with the return value of `get_completed_courses`. I'm not sure what you mean about unstarted items.

schenedx replied:
Thanks for the clarification. I didn't notice the `any` method used. Your explanation clarifies for me. Thanks!

jimabramson commented on openedx/core/djangoapps/programs/tests/test_utils.py:
```diff
@@ -10,20 +11,22 @@
+# This gets around an error about conflicting models documented at
+# https://code.djangoproject.com/ticket/22280. Importing the models from
+# `lms.djangoapps.certificates.models` raises a RuntimeError.
+from certificates.models import GeneratedCertificate  # pylint: disable=import-error
```
optional: not to be a stickler here (ok, maybe to be a stickler), we still have this reference to physical models from the certificates app.  It seems like it would be possible to expose the constants you're referencing via a simple variable or object in the API module, which would have the added benefit of avoiding the conflicting model workaround, but if it's not as simple as that, not a dealbreaker.

rlucioni replied:
@jimabramson good call, I've exposed the mode constants using a variable in the API module.

rlucioni commented:
jenkins run python

schenedx commented:
👍

jimabramson commented:
👍
